### PR TITLE
click on a switch to open/close it in svg

### PR DESCRIPTION
**Pull request by AbdelHedhili**
Signed-off-by: AbdelHedhili <abdelsalem.hedhili@rte-france.com>

### CODE REVIEW

File: src/components/study-pane.js [hunk 1]
```diff
@@ -175,6 +176,23 @@ const StudyPane = () => {
         history.replace("/studies/" + studyName)
     }
 
+    const handleUpdateSwitchState = useCallback( (studyName, breakerId, open, clickedDomElement, breakerMetadata) => {
+        updateSwitchState(studyName, breakerId, open).then( response => response.ok ? updateSvgSwitch(clickedDomElement, open, breakerMetadata) : console.debug(response));
+    }, []);
+
+    function updateSvgSwitch(clickedDomElement, open, breakerMetadata) {
```
Comment (geofjamg) [inline src/components/study-pane.js:183]: This is not the right way to do it as for instance when topological coloring will be activated, update can only be done on server side. So we need to reload again from the server the SVG generated with the new topology.

File: src/components/single-line-diagram.js [hunk 2]
```diff
@@ -77,9 +81,11 @@ const SingleLineDiagram = (props) => {
                     setSvg({svg: null, metadata: null, error, svgUrl: props.svgUrl});
                 });
         } else {
-            setSvg(nosvg);
+            setSvg(noSvg);
         }
-    }, [props.svgUrl]);
+    }, [props.svgUrl, useSelector(state => state.currentSvg)]);
```
Comment (jonenst) [inline src/components/single-line-diagram.js:86]: Hmm a dependency that is not used ? Is there no other way ?

File: src/components/single-line-diagram.js [hunk 3]
```diff
@@ -91,16 +97,20 @@ const SingleLineDiagram = (props) => {
...
             const svgHeight = bbox.height + 40;
-
             // using svgdotjs panzoom component to pan and zoom inside the svg, using svg width and height previously calculated for size and viewbox
             divElt.innerHTML = ""; // clear the previous svg in div element before replacing
             const draw = SVG()
                 .addTo(divElt)
                 .size(svgWidth, svgHeight)
                 .viewbox(xOrigin, yOrigin, svgWidth, svgHeight)
                 .panZoom({panning: true, zoomMin: 0.5, zoomMax: 10, zoomFactor: 0.3, margins: {top: svgHeight/4, left: svgWidth/4, bottom: svgHeight/4, right: svgWidth/4}});
+            console.log(svgDisplayInfo);
```
Comment (jonenst) [inline src/components/single-line-diagram.js:107]: remove debug log

**Comment by jonenst** [comment]
I don't think the state should be in the redux store, it's very local to the study-pane

File: src/components/single-line-diagram.js (hunk 2 above)
Comment (jonenst) [reply src/components/single-line-diagram.js:86]: Ah I see, then It's better to use a imperative handle to force the refresh of the svg I think

File: src/components/single-line-diagram.js [hunk 4]
```diff
@@ -77,9 +81,11 @@ const SingleLineDiagram = (props) => {
                     setSvg({svg: null, metadata: null, error, svgUrl: props.svgUrl});
                 });
         } else {
-            setSvg(nosvg);
+            setSvg(noSvg);
         }
-    }, [props.svgUrl]);
+    }, [props.svgUrl, useSelector(state => state.currentSvg)]);
+
+    const svgDisplayInfo = useSelector(state => state.svgDisplayInfo);
```
Comment (jonenst) [inline src/components/single-line-diagram.js:88]: should be a regular prop

File: src/components/study-pane.js [hunk 5]
```diff
@@ -230,6 +236,8 @@ const StudyPane = () => {
                             <div style={{ position: "absolute", left: 10, top: 10, zIndex: 1 }}>
                                 <SingleLineDiagram onClose={() => closeVoltageLevelDiagram()}
                                                    onNextVoltageLevelClick={showVoltageLevelDiagram}
+                                                   onBreakerClick={handleUpdateSwitchState}
+                                                   studyName={studyName}
```
Comment (jonenst) [inline src/components/study-pane.js:240]: should not need to pass the studyname by props

File: src/components/single-line-diagram.js [hunk 6]
```diff
@@ -60,11 +61,13 @@ const SvgNotFound = (props) => {
...
 
-const nosvg = {svg: null, metadata: null, error: null, svgUrl: null};
+const noSvg = {svg: null, metadata: null, error: null, svgUrl: null};
 
 const SingleLineDiagram = (props) => {
 
-    const [svg, setSvg] = useState(nosvg);
+    const [svg, setSvg] = useState(noSvg);
+
+    const studyName = useSelector(state => state.studyName);
```
Comment (jonenst) [inline src/components/single-line-diagram.js:70]: I don't think this component should know the study name. It already gets the full svg url containing the studyname from the parent component. Instead, it should expose a method in its ref (this is done with useImperativeHandle) to refresh the svg

File: src/components/single-line-diagram.js [hunk 7]
```diff
@@ -77,9 +80,11 @@ const SingleLineDiagram = (props) => {
                     setSvg({svg: null, metadata: null, error, svgUrl: props.svgUrl});
                 });
         } else {
-            setSvg(nosvg);
+            setSvg(noSvg);
         }
-    }, [props.svgUrl]);
```
Comment (jonenst) [inline src/components/single-line-diagram.js:82]: We shouldn't need to store the full svg just to test if it's the same as the previous one

File: src/components/study-pane.js [hunk 8]
```diff
@@ -175,6 +180,22 @@ const StudyPane = () => {
...
+                setCurrentSvg(fetchSvg(svgUrl));
+                setSvgDisplayInfo(svgDisplayInfo);
+            }
+            else {
+                console.error(response);
+            }
+        });
+    }, []);
+
+    const resetSvgDisplayInfo = useCallback(() => {
```
Comment (jonenst) [inline src/components/study-pane.js:195]: This should be replaced with the imperative ref

File: src/components/study-pane.js [hunk 9]
```diff
@@ -175,6 +180,22 @@ const StudyPane = () => {
         history.replace("/studies/" + studyName)
     }
 
+    const handleUpdateSwitchState = useCallback( (studyName, breakerId, open, svgUrl, svgDisplayInfo) => {
+        updateSwitchState(studyName, breakerId, open).then( response => {
+            if (response.ok) {
+                setCurrentSvg(fetchSvg(svgUrl));
+                setSvgDisplayInfo(svgDisplayInfo);
```
Comment (jonenst) [inline src/components/study-pane.js:187]: svgDisplayInfo should be local to the function that forces the refresh of an svg in the sld

File: src/components/single-line-diagram.js [hunk 10]
```diff
@@ -118,6 +130,21 @@ const SingleLineDiagram = (props) => {
                     const meta = svg.metadata.nodes.find( other => other.id === id );
                     props.onNextVoltageLevelClick(meta.nextVId);
                 })});
+
+            // handling the click on a breaker
+            const breakers = svg.metadata.nodes.filter(element => element.componentType === "BREAKER");
+            breakers.forEach(breaker => {
+                console.log(breaker);
```
Comment (jonenst) [inline src/components/single-line-diagram.js:137]: remove debug log

File: src/utils/rest-api.js [hunk 11]
```diff
@@ -117,6 +117,13 @@ export function deleteStudy(studyName) {
     });
 }
 
+export function updateSwitchState(studyName, switchId, open) {
+    console.info("updating switch " + switchId + " ...");
+    const updateSwitchUrl = process.env.REACT_APP_API_STUDY_SERVER + "/v1/studies/" + studyName + "/network-modification/switches/" + switchId + "?open=" + open;
+    console.debug(updateSwitchUrl);
+    return fetch(updateSwitchUrl, {method : 'put'});
```
Comment (jonenst) [inline src/utils/rest-api.js:124]: use (newly merged) backendFetch when merging master into this

File: src/components/study-pane.js [hunk 12]
```diff
@@ -89,6 +90,10 @@ const StudyPane = () => {
 
     const [filteredNominalVoltages, setFilteredNominalVoltages] = useState([]);
 
+    const [currentSvg, setCurrentSvg] = useState(null);
```
Comment (jonenst) [inline src/components/study-pane.js:93]: remove this ?

File: src/components/single-line-diagram.js [hunk 13]
```diff
@@ -60,11 +61,19 @@ const SvgNotFound = (props) => {
...
+const SingleLineDiagram = forwardRef((props, ref)  => {
 
-    const [svg, setSvg] = useState(nosvg);
+    const [svg, setSvg] = useState(noSvg);
+    const [svgDisplayInfo, setSvgDisplayInfo] = useState();
+
+    useImperativeHandle(ref, () => ({
+        reloadSvg: (svgUrl, svgDisplayInfo) => {
+            fetchSvg(svgUrl).then(svg => setSvg(svg));
+            setSvgDisplayInfo(svgDisplayInfo);
```
Comment (jonenst) [inline src/components/single-line-diagram.js:74]: i don't think svgDisplayInfo should be a state. instead it should be a ref that doesn't trigger relayouting when it is changed.

File: src/components/single-line-diagram.js [hunk 14]
```diff
@@ -60,11 +61,19 @@ const SvgNotFound = (props) => {
...
-const SingleLineDiagram = (props) => {
+const SingleLineDiagram = forwardRef((props, ref)  => {
 
-    const [svg, setSvg] = useState(nosvg);
+    const [svg, setSvg] = useState(noSvg);
+    const [svgDisplayInfo, setSvgDisplayInfo] = useState();
+
+    useImperativeHandle(ref, () => ({
+        reloadSvg: (svgUrl, svgDisplayInfo) => {
+            fetchSvg(svgUrl).then(svg => setSvg(svg));
```
Comment (jonenst) [inline src/components/single-line-diagram.js:73]: I don't think we should duplicate the fetchSvg call here. Instead we should just force a relayout of component:
```
  const [, updateState] = React.useState();
  const forceUpdate = React.useCallback(() => updateState({}), []);
```

File: src/components/single-line-diagram.js [hunk 15]
```diff
@@ -118,6 +130,21 @@ const SingleLineDiagram = (props) => {
...
+            breakers.forEach(breaker => {
+                console.log(breaker);
+                const domEl = document.getElementById(breaker.id);
+                domEl.style.cursor = "pointer";
+                domEl.addEventListener("click", function(event) {
+                    const clickedElementId = event.currentTarget.id;
+                    const breakerMetadata = svg.metadata.nodes.find(value => value.id === clickedElementId);
+                    const breakerId = breakerMetadata.equipmentId;
+                    const open = breakerMetadata.open;
+                    props.onBreakerClick(breakerId, !open, props.svgUrl, {viewBox :draw.viewbox(), zoom: draw.zoom()});
```
Comment (jonenst) [inline src/components/single-line-diagram.js:145]: shouldn't pass the viewbox and zoom to parent. Instead store them in a ref

Comment (jonenst) [reply src/components/single-line-diagram.js:145]: shouldn't pass the svgUrl to the parent

File: src/components/study-pane.js [hunk 16]
```diff
@@ -89,6 +90,10 @@ const StudyPane = () => {
...
+
+    const [svgDisplayInfo, setSvgDisplayInfo] = useState(null);
```
Comment (jonenst) [inline src/components/study-pane.js:95]: remove this ?

File: src/components/study-pane.js [hunk 17]
```diff
@@ -175,6 +180,17 @@ const StudyPane = () => {
...
+    const handleUpdateSwitchState = useCallback( (breakerId, open, svgUrl, svgDisplayInfo) => {
+        updateSwitchState(studyName, breakerId, open).then( response => {
+            if (response.ok) {
+                reloadSvg(svgUrl, svgDisplayInfo);
+            }
+            else {
+                console.error(response);
+            }
+        });
+    }, []);
```
Comment (jonenst) [inline src/components/study-pane.js:192]: missing studyName from the dependencies

File: src/components/study-pane.js [hunk 18]
```diff
@@ -199,6 +215,11 @@ const StudyPane = () => {
         mapRef.current.centerSubstation(network.getVoltageLevel(id).substationId);
     }, [mapRef, network]);
 
+    const sldRef = useRef();
+    const reloadSvg = useCallback((svgUrl, svgDisplayInfo)=> {
+        sldRef.current.reloadSvg(svgUrl, svgDisplayInfo);
+    }, [sldRef]);
```
Comment (jonenst) [inline src/components/study-pane.js:221]: I don't think you need sldRef in the deps, it doesn't change

**Comment by jonenst** [comment]
@geofjamg can you review again and mark your change requested as resolved if okay ?

**Pull request merged**